src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+            }
+            _ => Err(AlgorithmError::Unsupported),
+        }
+    }
+
+    /// Verify a signature.
+    pub fn verify(
+        &self,
+        signed_data: &[u8],
+        signature: &[u8],
+    ) -> Result<(), AlgorithmError> {
+        let valid = match self {
+            PublicKey::Default(digest_algorithm, public_key, _) => {
+                let mut verifier =
+                    Verifier::new(*digest_algorithm, public_key.as_ref())
+                        .map_err(|_| AlgorithmError::InvalidData)?;
+                verifier
+                    .verify_oneshot(signature, signed_data)
+                    .map_err(|_| AlgorithmError::InvalidData)?
+            }
+            PublicKey::NoDigest(public_key, _) => {
+                let mut verifier =
+                    Verifier::new_without_digest(public_key.as_ref())
+                        .map_err(|_| AlgorithmError::InvalidData)?;
+                verifier
+                    .verify_oneshot(signature, signed_data)
+                    .map_err(|_| AlgorithmError::InvalidData)?
+            }
+            PublicKey::EcDsa(digest_algorithm, public_key, _) => {
+                let half_len = signature.len() / 2;
+                let mut hasher = Hasher::new(*digest_algorithm)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                hasher
+                    .update(signed_data)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let hash = hasher
+                    .finish()
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let r = BigNum::from_slice(&signature[0..half_len])
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let s = BigNum::from_slice(&signature[half_len..])
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let ecdsa_sig = EcdsaSig::from_private_components(r, s)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                ecdsa_sig
+                    .verify(hash.as_ref(), public_key)
+                    .map_err(|_| AlgorithmError::InvalidData)?
+            }
+        };
+        if valid {
+            Ok(())
+        } else {
+            Err(AlgorithmError::BadSig)
+        }
+    }
+
+    /// Convert to a [`Dnskey`].
+    pub fn dnskey(&self) -> Dnskey<Vec<u8>> {
+        match self {
+            PublicKey::Default(message_digest, public_key, flags) => {
+                let alg = if *message_digest == MessageDigest::sha256() {
+                    SecAlg::RSASHA256
+                } else {
+                    unreachable!();
+                };
+                let rsa = public_key.rsa().expect("should not fail");
+                let e = rsa.e().to_vec();
+                let n = rsa.n().to_vec();
+
+                let mut key = Vec::new();
+
+                // Encode the exponent length.
+                if let Ok(exp_len) = u8::try_from(e.len()) {
```

Comment:
A reference to RFC 3110 section 2. "RSA Public KEY Resource Records" which defines the byte format of the key vec could be useful to readers here.